src/cpp/dds/domain/DomainParticipantImpl.cpp

```diff
@@ -0,0 +1,714 @@
+    subimpl->user_subscriber_ = sub;
+    subimpl->rtps_participant_ = this->rtps_participant_;
+
+    // Create InstanceHandle for the new subscriber
+    GUID_t sub_guid = guid();
+    do
+    {
+        sub_guid.entityId = fastrtps::rtps::c_EntityId_Unknown;
+        rtps_participant_->get_new_entity_id(sub_guid.entityId);
+    } while (exists_entity_id(sub_guid.entityId));
+
+    InstanceHandle_t sub_handle(sub_guid);
+    subimpl->handle_ = sub_handle;
+
+    //SAVE THE PUBLISHER INTO MAPS
+    std::lock_guard<std::mutex> lock(mtx_subs_);
+    subscribers_by_handle_[sub_handle] = sub;
+    subscribers_[sub] = subimpl;
+
+    return sub;
+}
+
+
+const TypeSupport DomainParticipantImpl::find_type(
+        const std::string& type_name) const
+{
+    std::lock_guard<std::mutex> lock(mtx_types_);
+
+    auto type_it = types_.find(type_name);
+
+    if (type_it != types_.end())
+    {
+        return type_it->second;
+    }
+
+    return nullptr;
+}
+
+bool DomainParticipantImpl::register_type(
+        const TypeSupport type,
+        const std::string& type_name)
+{
+    TypeSupport t = find_type(type_name);
+
+    if (t != nullptr)
+    {
+        if (t == type)
+        {
+            return true;
+        }
+
+        logError(PARTICIPANT, "Another type with the same name '" << type_name << "' is already registered.");
+    }
+
+    if (type_name.size() <= 0)
+    {
+        logError(PARTICIPANT, "Registered Type must have a name");
+        return false;
+    }
+
+    logInfo(PARTICIPANT, "Type " << type_name << " registered.");
+    std::lock_guard<std::mutex> lock(mtx_types_);
+    types_.insert(std::make_pair(type_name, type));
+    return true;
+}
+
+bool DomainParticipantImpl::unregister_type(
+        const char* type_name)
+{
+    TypeSupport t = find_type(type_name);
+
+    if (t.empty())
+    {
+        return true; // Not registered, so unregistering complete.
+    }
+
+    {
+        // Check is any subscriber is using the type
+        std::lock_guard<std::mutex> lock(mtx_subs_);
+
+        for (auto sit : subscribers_)
+        {
+            std::vector<DataReader*> readers;
+            sit.second->get_datareaders(readers);
+            for (DataReader* reader : readers)
+            {
+                if (reader->get_topic().getTopicDataType() == type_name)
+                {
+                    return false; // Is in use
+                }
+            }
+        }
+    }
+
+    {
+        // Check is any publisher is using the type
+        std::lock_guard<std::mutex> lock(mtx_pubs_);
+
+        for (auto pit : publishers_)
+        {
+            std::vector<DataWriter*> writers;
+            pit.second->get_datawriters(writers);
+            for (DataWriter* writer : writers)
+            {
+                if (writer->get_topic().getTopicDataType() == type_name)
+                {
+                    return false; // Is in use
+                }
+            }
+        }
+    }
+
+    types_.erase(type_name);
```

Comment:
Protect this access with a mutex
```suggestion
    std::lock_guard<std::mutex> lock(mtx_types_);
    types_.erase(type_name);
```

Reply:
Good catch!